### DISCUSSION TITLE
Did you test your setup (NanoPi Neo) with Windows 10 Host or Linux?

### DISCUSSION THREAD

**Discussion by xbing6** [parkrrrr/bdio#13] [Q&A]
This is awesome project. One question: did you test your setup (NanoPi Neo) with Windows 10 Host or Linux like Ubuntu?

I understand you connected HIMS SyncBraille to NanoPi Neo.

**Comment by parkrrrr** [comment]
I'm not exactly sure what question you're asking, but I'll try to answer.

The NanoPi Neo is sort of an intermediary in this project. It acts as a USB host for the SyncBraille, and as a USB device for another computer. The SyncBraille internally uses an FTDI FT232 USB-to-serial chip, so it looks like a serial port if you plug it directly into your host computer. Putting the NanoPi running this software in the middle makes it look like a modern Braille display that speaks HUTRR-78. Currently, there's little to no advantage to this, but the vision for the future is that you'll be able to just plug it in and your screen reader will automatically recognize it and know how many cells and what buttons it has. 

**The NanoPi runs a custom build of OpenWRT, which is based on Linux** and designed for smaller embedded systems like wifi routers and such. Once the software is installed, ideally, you never have to worry about what's running on the Neo, because it just works. (It's not to that point, yet, because I haven't been able to find the time to do any development work on it since the pandemic started.)

**The other computer, the one that you would be interacting with in normal use, runs whatever OS you like.** You plug the Braille display into the USB-A port on the Neo, and you plug a cable into the USB-OTG port on the Neo and the USB-A port on your desktop or laptop computer, and as far as the desktop computer knows it's plugged directly into a Braille display that supports HID Braille. In my case, **all of my tests have been with a Windows 10 host** running custom HID Braille software that I can't release, but the host is mostly irrelevant as long as the software knows how to work with HUTRR-78/HID Braille.

Note that as far as I know, there is little to no support in current screen readers for HID Braille - that's one of the reasons I started this project: to try to break the logjam of "there's no software support so why create hardware / there's no hardware so why add this feature to our software" that's been pervading the AT market. Also, I don't know if screen readers that do support HID Braille will work properly if they get attached to a device they weren't expecting; it all depends on how they implement their drivers. Current screen readers have databases of what keys are available on a given device and what mappings to use for them, for example, so they might attempt to do a database lookup with the manufacturer/model strings and fail if they're not found. We'll all just have to wait and see how that shakes out.

**Comment by parkrrrr** [comment]
(Note: I'm going to attempt to convert this issue to a discussion, since it's not an action item, so it may look like it's been deleted from the Issues view.)

**Comment by xbing6** [comment]
Ron, thanks a lot for your very quick response.

=>  In my case, all of my tests have been with a Windows 10 host running custom HID Braille software that I can't release, but the host is mostly irrelevant as long as the software knows how to work with HUTRR-78/HID Braille.

This is exactly I am looking for. But, sadly this means that there is no way I can replicate your test setup (Windows 10 Host + un-released HID Braille software <--> NanoPi Neo as an intermediary <--> SyncBraille (internally functions as FTDI FT232 USB-to-serial), because I do not have HID Braille software that you can *not* release.

I sent an email to BRLTTY mailing list (link below) to see if BRLTTY supports HUTRR-78/HID Braille, from BRLTTY "git commits history" I can not find any information relating to HUTRR-78. Even if BRLTTY does *not* support HUTRR-78 *today*, I do envision that BRLTTY will support HUTRR-78 later, and there is a chance that we may be able to use HUTRR-78 on Ubuntu / Linux Host (with BRLTTY) later.
https://brltty.app/pipermail/brltty/2020-December/017927.html

Thanks again.

**Comment by parkrrrr** [comment]
You might be amused to hear that what's running on the NanoPi *is* BRLTTY, or a very heavily modified build of it, anyway, along with some other software to do the translation to HID Braille. 

The way BRLTTY drivers work, it might be difficult to add HID Braille support to it. I haven't investigated how difficult it would be to build their various descriptor tables dynamically, but it would definitely be different from how everything else is described at present. If I had more free time, I'd give it a shot, since I think I understand HID Braille as well as just about anyone at this point, but I'd like to get this one working with other displays first.

Unfortunately, my employer owns the rights to the other software I've been using, so I can't really say anything about it.

**Comment by xbing6** [comment]
=> You might be amused to hear that what's running on the NanoPi is BRLTTY,

I understand this, as you clearly write it in README: BDIOHID runs on a small Linux Single-Board Computer and **leverages brltty to communicate with legacy hardware**, 

BRLTTY can play different roles:
- Running on NanoPi Neo, to communicate with legacy hardware
- Running on Ubuntu / Linux Host
- (I am not exactly sure) Microsoft Narrator (on Windows 10) seemingly uses BRLTTY or liblouis. If Microsoft Narrator does leverage BRLTTY or liblouis, I can understand it too.

BRLTTY guys replied my question about if it supports HUTRR-78, the answer is that BRLTTY does not support it **today**
https://brltty.app/pipermail/brltty/2020-December/017933.html

Besides, Dave Mielke who answered my question is the maintainer of BRLTTY.
http://mielke.cc/brltty/contact.html
